File: src/FilesToData.Api/Services/SupabaseService.cs
using System.Collections.Concurrent;
using System.Text.Json;
using FilesToData.Shared.Models;
using Microsoft.Extensions.Logging;

namespace FilesToData.Api.Services;

/// <summary>
/// Implementación en memoria de ISupabaseService.
/// No persiste en Postgres real, pero permite probar el flujo completo localmente.
/// </summary>
public class SupabaseService : ISupabaseService
{
    private readonly ILogger<SupabaseService> _logger;

    private static readonly ConcurrentDictionary<Guid, Job> _jobs = new();
    private static readonly ConcurrentDictionary<Guid, object> _results = new();
    private static readonly ConcurrentDictionary<Guid, List<MaskingLog>> _maskingLogs = new();
    private static readonly ConcurrentDictionary<string, byte[]> _files = new();

    public SupabaseService(ILogger<SupabaseService> logger)
    {
        _logger = logger;
    }

    // =========================================================================
    // JOBS (in-memory)
    // =========================================================================

    public Task<Job> CreateJobAsync(Job job)
    {
        _jobs[job.Id] = job;
        _logger.LogInformation("Job {JobId} creado en memoria", job.Id);
        return Task.FromResult(job);
    }

    public Task<Job?> GetJobAsync(Guid jobId)
    {
        _jobs.TryGetValue(jobId, out var job);
        return Task.FromResult<Job?>(job);
    }

    public Task<List<Job>> ListJobsAsync(JobStatus? status = null, JobMode? mode = null, int limit = 50, int offset = 0)
    {
        var query = _jobs.Values.AsQueryable();

        if (status.HasValue)
            query = query.Where(j => j.Status == status.Value);

        if (mode.HasValue)
            query = query.Where(j => j.Mode == mode.Value);

        var list = query
            .OrderByDescending(j => j.CreatedAt)
            .Skip(offset)
            .Take(limit)
            .ToList();

        return Task.FromResult(list);
    }

    public Task UpdateJobStatusAsync(Guid jobId, JobStatus status, string? errorMessage = null)
    {
        if (_jobs.TryGetValue(jobId, out var job))
        {
            job.Status = status;
            job.ErrorMessage = errorMessage;
            job.UpdatedAt = DateTime.UtcNow;
            _jobs[jobId] = job;
        }

        _logger.LogInformation("Job {JobId} actualizado a {Status}", jobId, status);
        return Task.CompletedTask;
    }

    // =========================================================================
    // RESULTS
    // =========================================================================

    public Task<object?> GetResultsAsync(Guid jobId)
    {
        _results.TryGetValue(jobId, out var data);
        return Task.FromResult(data);
    }

    public Task UpsertResultsAsync(Guid jobId, object data)
    {
        _results[jobId] = data;
        _logger.LogInformation("Resultados actualizados para job {JobId}", jobId);
        return Task.CompletedTask;
    }

    // =========================================================================
    // MASKING LOGS
    // =========================================================================

    public Task CreateMaskingLogAsync(MaskingLog log)
    {
        var list = _maskingLogs.GetOrAdd(log.JobId, _ => new List<MaskingLog>());
        list.Add(log);
        _logger.LogInformation("MaskingLog agregado para job {JobId}", log.JobId);
        return Task.CompletedTask;
    }

    public Task<List<MaskingLog>> GetMaskingLogsAsync(Guid jobId)
    {
        if (_maskingLogs.TryGetValue(jobId, out var list))
        {
            return Task.FromResult(list.ToList());
        }

        return Task.FromResult(new List<MaskingLog>());
    }

    // =========================================================================
    // STORAGE
    // =========================================================================

    public Task UploadFileAsync(string filePath, byte[] content)
    {
        _files[filePath] = content;
        _logger.LogInformation("Archivo {Path} almacenado en memoria", filePath);
        return Task.CompletedTask;
    }

    public Task<byte[]> DownloadFileAsync(string filePath)
    {
        if (_files.TryGetValue(filePath, out var content))
        {
            return Task.FromResult(content);
        }

        throw new FileNotFoundException("File not found in in-memory storage", filePath);
    }

    public Task<string> GetFileUrlAsync(string filePath, int expiresIn = 3600)
    {
        // En modo memoria devolvemos solo la ruta
        return Task.FromResult($"memory://{filePath}");
    }
}

File: src/FilesToData.ApiPy/services/supabase_service.py
import threading
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


class SupabaseService:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._jobs: Dict[str, Dict[str, Any]] = {}
        self._results: Dict[str, Any] = {}
        self._masking_logs: Dict[str, List[Dict[str, Any]]] = {}
        self._files: Dict[str, bytes] = {}

    # Jobs
    def create_job(self, job: Dict[str, Any]) -> Dict[str, Any]:
        with self._lock:
            self._jobs[job["id"]] = job
        return job

    def get_job(self, job_id: str) -> Optional[Dict[str, Any]]:
        with self._lock:
            return self._jobs.get(job_id)

    def list_jobs(
        self,
        status: Optional[str] = None,
        mode: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Dict[str, Any]]:
        with self._lock:
            jobs = list(self._jobs.values())

        if status:
            jobs = [j for j in jobs if str(j.get("status", "")).upper() == status.upper()]
        if mode:
            jobs = [j for j in jobs if str(j.get("mode", "")).upper() == mode.upper()]

        jobs.sort(key=lambda j: j.get("created_at") or "", reverse=True)
        return jobs[offset : offset + limit]

    def update_job_status(self, job_id: str, status: str, error_message: Optional[str] = None) -> None:
        with self._lock:
            job = self._jobs.get(job_id)
            if not job:
                return
            job["status"] = status
            job["error_message"] = error_message
            job["updated_at"] = _utc_now_iso()
            self._jobs[job_id] = job

    # Results
    def get_results(self, job_id: str) -> Any:
        with self._lock:
            return self._results.get(job_id)

    def upsert_results(self, job_id: str, data: Any) -> None:
        with self._lock:
            self._results[job_id] = data

    # Masking logs
    def create_masking_log(self, log: Dict[str, Any]) -> None:
        job_id = str(log.get("job_id"))
        with self._lock:
            self._masking_logs.setdefault(job_id, []).append(log)

    def get_masking_logs(self, job_id: str) -> List[Dict[str, Any]]:
        with self._lock:
            return list(self._masking_logs.get(job_id, []))

    # Storage
    def upload_file(self, file_path: str, content: bytes) -> None:
        with self._lock:
            self._files[file_path] = content

    def download_file(self, file_path: str) -> bytes:
        with self._lock:
            if file_path not in self._files:
                raise FileNotFoundError(f"File not found in in-memory storage: {file_path}")
            return self._files[file_path]


supabase = SupabaseService()
